composer/datasets/synthetic.py

```diff
@@ -21,6 +36,201 @@ class SyntheticDataLabelType(StringEnum):
+        from transformers import BertTokenizer
+    except ImportError as e:
+        raise ImportError(
+            textwrap.dedent("""\
+            Composer was installed without NLP support. To use NLP with Composer, run `pip install mosaicml[nlp]`
+            if using pip or `conda install -c conda-forge transformers tokenizers` if using Anaconda.""")) from e
+
+    unk_token = "[UNK]"
+    pad_token = "[PAD]"
+
+    initial_alphabet = "".join([i for i in dataset])
+    initial_alphabet = list(set(initial_alphabet))
+
+    return SyntheticTokenizerParams(
+        tokenizer_model=tokenizers_models.WordPiece(unk_token=unk_token),  # type: ignore
+        normalizer=normalizers.BertNormalizer(),
+        pre_tokenizer=pre_tokenizers.BertPreTokenizer(),
+        decoder=decoders.WordPiece(),
+        initial_alphabet=initial_alphabet,
+        special_tokens=[pad_token, unk_token, "[SEP]", "[CLS]", "[MASK]"],
+        pad_token=pad_token,
+        trainer_cls=tokenizers_trainer.WordPieceTrainer,
+        tokenizer_cls=BertTokenizer,
+    )
+
+
+def generate_gpt2_tokenizer_params() -> SyntheticTokenizerParams:
+    try:
+        import tokenizers.decoders as decoders
+        import tokenizers.models as tokenizers_models
+        import tokenizers.normalizers as normalizers
+        import tokenizers.pre_tokenizers as pre_tokenizers
+        import tokenizers.trainers as tokenizers_trainer
+        from transformers import GPT2Tokenizer
+    except ImportError as e:
+        raise ImportError(
+            textwrap.dedent("""\
+            Composer was installed without NLP support. To use NLP with Composer, run `pip install mosaicml[nlp]`
+            if using pip or `conda install -c conda-forge tokenizers` if using Anaconda.""")) from e
+
+    unk_token = None
+    pad_token = "<pad>"
+
+    return SyntheticTokenizerParams(
+        tokenizer_model=tokenizers_models.BPE(unk_token=unk_token),
+        normalizer=normalizers.Lowercase(),
+        pre_tokenizer=pre_tokenizers.ByteLevel(),
+        decoder=decoders.ByteLevel(),
+        initial_alphabet=pre_tokenizers.ByteLevel.alphabet(),
+        special_tokens=[pad_token, "<endoftext>"],
+        pad_token=pad_token,
+        trainer_cls=tokenizers_trainer.BpeTrainer,
+        tokenizer_cls=GPT2Tokenizer,
+    )
+
+
+def generate_synthetic_tokenizer(tokenizer_family: str, dataset=None, vocab_size=256) -> PreTrainedTokenizer:
+    try:
+        from tokenizers import Tokenizer
+        from transformers import PreTrainedTokenizer
+    except ImportError as e:
+        raise ImportError(
+            textwrap.dedent("""\
+            Composer was installed without NLP support. To use NLP with Composer, run `pip install mosaicml[nlp]`
+            if using pip or `conda install -c conda-forge transformers tokenizers` if using Anaconda.""")) from e
+
+    # generate a synthetic dataset with reasonable defaults is none is provided
+    if dataset is None:
+        num_samples = 100
```

Comment:
should these be parameters?

Reply:
In my opinion, if the user wants to change these parameters, they should pass in a dataset themselves. We could make these into parameters, and pass that into the dataset (to the same effect), but I thought this would enable a short function signature with the same flexibility. In other words, those are just defaults